--- a/scripts/gcc-plugins/Makefile	2023-02-02 14:09:24.615360391 -0500
+++ b/scripts/gcc-plugins/Makefile	2023-02-02 14:09:51.422139879 -0500
@@ -22,7 +22,7 @@ always-y += $(GCC_PLUGIN)
 GCC_PLUGINS_DIR = $(shell $(CC) -print-file-name=plugin)
 
 plugin_cxxflags	= -Wp,-MMD,$(depfile) $(KBUILD_HOSTCXXFLAGS) -fPIC \
-		   -I $(GCC_PLUGINS_DIR)/include -I $(obj) -std=gnu++11 \
+			 -I $(GCC_PLUGINS_DIR)/include -I $(obj) \
 		   -fno-rtti -fno-exceptions -fasynchronous-unwind-tables \
 		   -ggdb -Wno-narrowing -Wno-unused-variable \
 		   -Wno-format-diag
