diff --git a/Kconfig b/Kconfig
index 745bc773f..e306bacea 100644
--- a/Kconfig
+++ b/Kconfig
@@ -30,3 +30,5 @@ source "lib/Kconfig"
 source "lib/Kconfig.debug"
 
 source "Documentation/Kconfig"
+
+source "distro/Kconfig"
diff --git a/distro/Kconfig b/distro/Kconfig
new file mode 100644
index 000000000..94d6e1886
--- /dev/null
+++ b/distro/Kconfig
@@ -0,0 +1,295 @@
+menu "Gentoo Linux"
+
+config GENTOO_LINUX
+	bool "Gentoo Linux support"
+
+	default y
+
+	help
+		In order to boot Gentoo Linux a minimal set of config settings needs to
+		be enabled in the kernel; to avoid the users from having to enable them
+		manually as part of a Gentoo Linux installation or a new clean config,
+		we enable these config settings by default for convenience.
+
+		See the settings that become available for more details and fine-tuning.
+
+config GENTOO_LINUX_UDEV
+	bool "Linux dynamic and persistent device naming (userspace devfs) support"
+
+	depends on GENTOO_LINUX
+	default y if GENTOO_LINUX
+
+	select DEVTMPFS
+	select TMPFS
+	select UNIX
+
+	select MMU
+	select SHMEM
+
+	help
+		In order to boot Gentoo Linux a minimal set of config settings needs to
+		be enabled in the kernel; to avoid the users from having to enable them
+		manually as part of a Gentoo Linux installation or a new clean config,
+		we enable these config settings by default for convenience.
+
+		Currently this only selects TMPFS, DEVTMPFS and their dependencies.
+		TMPFS is enabled to maintain a tmpfs file system at /dev/shm, /run and
+		/sys/fs/cgroup; DEVTMPFS to maintain a devtmpfs file system at /dev.
+
+		Some of these are critical files that need to be available early in the
+		boot process; if not available, it causes sysfs and udev to malfunction.
+
+		To ensure Gentoo Linux boots, it is best to leave this setting enabled;
+		if you run a custom setup, you could consider whether to disable this.
+
+config GENTOO_LINUX_PORTAGE
+	bool "Select options required by Portage features"
+
+	depends on GENTOO_LINUX
+	default y if GENTOO_LINUX
+
+	select CGROUPS
+	select NAMESPACES
+	select IPC_NS
+	select NET_NS
+	select PID_NS
+	select SYSVIPC
+	select USER_NS
+	select UTS_NS
+
+	help
+		This enables options required by various Portage FEATURES.
+		Currently this selects:
+
+		CGROUPS     (required for FEATURES=cgroup)
+		IPC_NS      (required for FEATURES=ipc-sandbox)
+		NET_NS      (required for FEATURES=network-sandbox)
+		PID_NS      (required for FEATURES=pid-sandbox)
+		SYSVIPC     (required by IPC_NS)
+
+		It is highly recommended that you leave this enabled as these FEATURES
+		are, or will soon be, enabled by default.
+
+menu "Support for init systems, system and service managers"
+	visible if GENTOO_LINUX
+
+config GENTOO_LINUX_INIT_SCRIPT
+	bool "OpenRC, runit and other script based systems and managers"
+
+	default y if GENTOO_LINUX
+
+	depends on GENTOO_LINUX
+
+	select BINFMT_SCRIPT
+	select CGROUPS
+	select EPOLL
+	select FILE_LOCKING
+	select INOTIFY_USER
+	select SIGNALFD
+	select TIMERFD
+
+	help
+		The init system is the first thing that loads after the kernel booted.
+
+		These config settings allow you to select which init systems to support;
+		instead of having to select all the individual settings all over the
+		place, these settings allows you to select all the settings at once.
+
+		This particular setting enables all the known requirements for OpenRC,
+		runit and similar script based systems and managers.
+
+		If you are unsure about this, it is best to leave this setting enabled.
+
+config GENTOO_LINUX_INIT_SYSTEMD
+	bool "systemd"
+
+	default n
+
+	depends on GENTOO_LINUX && GENTOO_LINUX_UDEV
+
+	select AUTOFS4_FS
+	select BLK_DEV_BSG
+	select BPF_SYSCALL
+	select CGROUP_BPF
+	select CGROUPS
+	select CRYPTO_HMAC
+	select CRYPTO_SHA256
+	select CRYPTO_USER_API_HASH
+	select DEVPTS_MULTIPLE_INSTANCES
+	select DMIID if X86_32 || X86_64 || X86
+	select EPOLL
+	select FANOTIFY
+	select FHANDLE
+	select FILE_LOCKING
+	select INOTIFY_USER
+	select IPV6
+	select KCMP
+	select NET
+	select NET_NS
+	select PROC_FS
+	select SECCOMP if HAVE_ARCH_SECCOMP
+	select SECCOMP_FILTER if HAVE_ARCH_SECCOMP_FILTER
+	select SIGNALFD
+	select SYSFS
+	select TIMERFD
+	select TMPFS_POSIX_ACL
+	select TMPFS_XATTR
+
+	select ANON_INODES
+	select BLOCK
+	select EVENTFD
+	select FSNOTIFY
+	select INET
+	select NLATTR
+
+	help
+		The init system is the first thing that loads after the kernel booted.
+
+		These config settings allow you to select which init systems to support;
+		instead of having to select all the individual settings all over the
+		place, these settings allows you to select all the settings at once.
+
+		This particular setting enables all the known requirements for systemd;
+		it also enables suggested optional settings, as the package suggests to.
+
+endmenu
+
+menu "Kernel Self Protection Project"
+	visible if GENTOO_LINUX
+
+config GENTOO_KERNEL_SELF_PROTECTION
+	bool "Enable Kernel Self Protection Project Recommendations"
+
+	depends on GENTOO_LINUX && EXPERT && !DEVKMEM && !PROC_KCORE && !COMPAT_VDSO && !KEXEC && !HIBERNATION && !LEGACY_PTYS && !MODIFY_LDT_SYSCALL
+
+	select BUG
+	select STRICT_KERNEL_RWX if ARCH_HAS_STRICT_KERNEL_RWX
+	select DEBUG_FS
+	select DEBUG_WX if ARCH_HAS_DEBUG_WX && MMU
+	select STACKPROTECTOR if HAVE_STACKPROTECTOR
+	select STACKPROTECTOR_STRONG if HAVE_STACKPROTECTOR
+	select STRICT_DEVMEM if DEVMEM=y && (ARCH_HAS_DEVMEM_IS_ALLOWED || GENERIC_LIB_DEVMEM_IS_ALLOWED)
+	select IO_STRICT_DEVMEM if STRICT_DEVMEM
+	select SYN_COOKIES if NET && INET
+	select DEBUG_CREDENTIALS if DEBUG_KERNEL
+	select DEBUG_NOTIFIERS if DEBUG_KERNEL
+	select DEBUG_LIST
+	select DEBUG_SG if DEBUG_KERNEL
+	select BUG_ON_DATA_CORRUPTION
+	select SCHED_STACK_END_CHECK if DEBUG_KERNEL
+	select SECCOMP if HAVE_ARCH_SECCOMP
+	select SECCOMP_FILTER if HAVE_ARCH_SECCOMP_FILTER
+	select SECURITY if SYSFS && MULTIUSER
+	select SECURITY_YAMA if SECURITY
+	select HARDENED_USERCOPY if HAVE_HARDENED_USERCOPY_ALLOCATOR
+	select SLAB_FREELIST_RANDOM if SLAB || SLUB
+	select SLAB_FREELIST_HARDENED if SLAB || SLUB
+	select SHUFFLE_PAGE_ALLOCATOR
+	select SLUB_DEBUG if SLUB && SYSFS
+	select SLUB_DEBUG_ON if SLUB_DEBUG
+	select PAGE_POISONING
+	select PAGE_POISONING_NO_SANITY
+	select PAGE_POISONING_ZERO
+	select INIT_ON_ALLOC_DEFAULT_ON
+	select INIT_ON_FREE_DEFAULT_ON
+	select FORTIFY_SOURCE if ARCH_HAS_FORTIFY_SOURCE && !CC_IS_CLANG
+	select SECURITY_DMESG_RESTRICT 
+	select PANIC_ON_OOPS
+	select DEBUG_STACKOVERFLOW if DEBUG_KERNEL && HAVE_DEBUG_STACKOVERFLOW
+	select VMAP_STACK if HAVE_ARCH_VMAP_STACK
+	select STRICT_MODULE_RWX if ARCH_HAS_STRICT_MODULE_RWX && ARCH_OPTIONAL_KERNEL_RWX && MODULES
+	select ZERO_CALL_USED_REGS if CC_HAS_ZERO_CALL_USED_REGS
+	select INIT_STACK_ALL_PATTERN if CC_HAS_AUTO_VAR_INIT_PATTERN && !CC_HAS_AUTO_VAR_INIT_ZERO
+	select INIT_STACK_ALL_ZERO if CC_HAS_AUTO_VAR_INIT_ZERO
+	select GCC_PLUGINS if HAVE_GCC_PLUGINS && CC_IS_GCC
+	select GCC_PLUGIN_LATENT_ENTROPY if GCC_PLUGINS
+	select GCC_PLUGIN_STRUCTLEAK if GCC_PLUGINS
+	select GCC_PLUGIN_STRUCTLEAK_BYREF_ALL if GCC_PLUGINS
+	select GCC_PLUGIN_STRUCTLEAK_VERBOSE if GCC_PLUGINS && GCC_PLUGIN_STRUCTLEAK
+	select GCC_PLUGIN_RANDSTRUCT if GCC_PLUGINS
+	select GCC_PLUGIN_RANDSTRUCT_PERFORMANCE if GCC_PLUGINS && GCC_PLUGIN_RANDSTRUCT
+	select GCC_PLUGIN_STACKLEAK if GCC_PLUGINS && HAVE_ARCH_STACKLEAK
+
+	help
+		Recommended Kernel settings based on the suggestions from the Kernel Self Protection Project
+		See: https://kernsec.org/wiki/index.php/Kernel_Self_Protection_Project/Recommended_Settings
+		Note, there may be additional settings for which the CONFIG_ setting is invisible in menuconfig due
+		to unmet dependencies. Search for GENTOO_KERNEL_SELF_PROTECTION and search for
+		GENTOO_KERNEL_SELF_PROTECTION_{X86_64, ARM64, X86_32, ARM} for dependency information on your
+		specific architecture.
+		Note 2: Please see the URL above for numeric settings, e.g. CONFIG_DEFAULT_MMAP_MIN_ADDR=65536
+		for X86_64
+
+config GENTOO_KERNEL_SELF_PROTECTION_X86_64
+	bool "X86_64 KSPP Settings" if GENTOO_KERNEL_SELF_PROTECTION
+
+	depends on X86_64 && GENTOO_KERNEL_SELF_PROTECTION
+	default y if X86_64 && GENTOO_KERNEL_SELF_PROTECTION
+
+	select RANDOMIZE_BASE
+	select RANDOMIZE_MEMORY
+	select RELOCATABLE
+	select LEGACY_VSYSCALL_NONE
+	select PAGE_TABLE_ISOLATION
+
+
+config GENTOO_KERNEL_SELF_PROTECTION_ARM64
+	bool "ARM64 KSPP Settings"
+
+	depends on ARM64 && GENTOO_KERNEL_SELF_PROTECTION
+	default y if ARM64 && GENTOO_KERNEL_SELF_PROTECTION
+
+	select RANDOMIZE_BASE
+	select RELOCATABLE
+	select ARM64_SW_TTBR0_PAN
+	select CONFIG_UNMAP_KERNEL_AT_EL0
+
+config GENTOO_KERNEL_SELF_PROTECTION_X86_32
+	bool "X86_32 KSPP Settings"
+
+	depends on !X86_MSR && !MODIFY_LDT_SYSCALL && !M486 && X86_32 && GENTOO_KERNEL_SELF_PROTECTION
+	default y if X86_32 && GENTOO_KERNEL_SELF_PROTECTION
+
+	#select HIGHMEM64G
+	#select X86_PAE
+	select RANDOMIZE_BASE
+	select RELOCATABLE
+	select PAGE_TABLE_ISOLATION
+
+config GENTOO_KERNEL_SELF_PROTECTION_ARM
+	bool "ARM KSPP Settings"
+
+	depends on !OABI_COMPAT && ARM && GENTOO_KERNEL_SELF_PROTECTION
+	default y if ARM && GENTOO_KERNEL_SELF_PROTECTION
+
+	select VMSPLIT_3G
+	select STRICT_MEMORY_RWX
+	select CPU_SW_DOMAIN_PAN
+
+config GENTOO_KERNEL_SELF_PROTECTION_PPC
+	bool "PPC KSPP Settings"
+
+	depends on !SCOM_DEBUGFS && !OPAL_CORE && PPC && GENTOO_KERNEL_SELF_PROTECTION
+	default y if PPC && GENTOO_KERNEL_SELF_PROTECTION
+
+	select PPC_KUEP if PPC_HAVE_KUEP
+	select PPC_KUAP if PPC_HAVE_KUAP
+	select PPC_MEM_KEYS if PPC_BOOK3S_64
+	select PPC_SUBPAGE_PROT if PPC_BOOK3S_64 && PPC_64K_PAGES
+
+endmenu
+
+config GENTOO_PRINT_FIRMWARE_INFO
+	bool "Print firmware information that the kernel attempts to load"
+
+	depends on GENTOO_LINUX
+	default y
+
+	help
+		Enable this option to print information about firmware that the kernel
+		is attempting to load.  This information can be accessible via the
+		dmesg command-line utility
+
+		See the settings that become available for more details and fine-tuning.
+
+endmenu
diff --git a/drivers/acpi/Kconfig b/drivers/acpi/Kconfig
index 1da360c51..70963ba91 100644
--- a/drivers/acpi/Kconfig
+++ b/drivers/acpi/Kconfig
@@ -445,7 +445,7 @@ config ACPI_HED
 
 config ACPI_CUSTOM_METHOD
 	tristate "Allow ACPI methods to be inserted/replaced at run time"
-	depends on DEBUG_FS
+	depends on DEBUG_FS && !GENTOO_KERNEL_SELF_PROTECTION
 	help
 	  This debug facility allows ACPI AML methods to be inserted and/or
 	  replaced without rebooting the system. For details refer to:
diff --git a/init/Kconfig b/init/Kconfig
index 11f8a845f..9f3eff46f 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1879,6 +1879,7 @@ config SLUB_DEBUG
 config COMPAT_BRK
 	bool "Disable heap randomization"
 	default y
+	depends on !GENTOO_KERNEL_SELF_PROTECTION
 	help
 	  Randomizing heap placement makes heap exploits harder, but it
 	  also breaks ancient binaries (including anything libc5 based).
@@ -1925,7 +1926,9 @@ endchoice
 
 config SLAB_MERGE_DEFAULT
 	bool "Allow slab caches to be merged"
+	default n if GENTOO_KERNEL_SELF_PROTECTION
 	default y
+	depends on !GENTOO_KERNEL_SELF_PROTECTION
 	help
 	  For reduced kernel memory fragmentation, slab caches can be
 	  merged when they share the same size and other characteristics.
diff --git a/mm/Kconfig b/mm/Kconfig
index c048dea7e..81a1dfd69 100644
--- a/mm/Kconfig
+++ b/mm/Kconfig
@@ -305,6 +305,8 @@ config KSM
 config DEFAULT_MMAP_MIN_ADDR
 	int "Low address space to protect from user allocation"
 	depends on MMU
+	default 65536 if ( X86_64 || X86_32 || PPC64 || IA64 ) && GENTOO_KERNEL_SELF_PROTECTION
+	default 32768 if ( ARM64 || ARM ) && GENTOO_KERNEL_SELF_PROTECTION
 	default 4096
 	help
 	  This is the portion of low virtual memory which should be protected
diff --git a/security/Kconfig.hardening b/security/Kconfig.hardening
index 90cbaff86..7b48339e8 100644
--- a/security/Kconfig.hardening
+++ b/security/Kconfig.hardening
@@ -30,6 +30,7 @@ choice
 	default GCC_PLUGIN_STRUCTLEAK_BYREF_ALL if COMPILE_TEST && GCC_PLUGINS
 	default INIT_STACK_ALL_PATTERN if COMPILE_TEST && CC_HAS_AUTO_VAR_INIT_PATTERN
 	default INIT_STACK_ALL_ZERO if CC_HAS_AUTO_VAR_INIT_PATTERN
+	default GCC_PLUGIN_STRUCTLEAK_BYREF_ALL if GENTOO_KERNEL_SELF_PROTECTION && GCC_PLUGINS
 	default INIT_STACK_NONE
 	help
 	  This option enables initialization of stack variables at
@@ -45,6 +46,7 @@ choice
 
 	config INIT_STACK_NONE
 		bool "no automatic stack variable initialization (weakest)"
+		depends on !GENTOO_KERNEL_SELF_PROTECTION
 		help
 		  Disable automatic stack variable initialization.
 		  This leaves the kernel vulnerable to the standard
diff --git a/security/selinux/Kconfig b/security/selinux/Kconfig
index 9e921fc72..f29bc13fa 100644
--- a/security/selinux/Kconfig
+++ b/security/selinux/Kconfig
@@ -26,6 +26,7 @@ config SECURITY_SELINUX_BOOTPARAM
 config SECURITY_SELINUX_DISABLE
 	bool "NSA SELinux runtime disable"
 	depends on SECURITY_SELINUX
+	depends on !GENTOO_KERNEL_SELF_PROTECTION
 	select SECURITY_WRITABLE_HOOKS
 	default n
 	help
